### ISSUE TITLE
O(n^2) behavior inside `consume_markers`

### ISSUE THREAD

**Issue by johnstiles-google** [libjpeg-turbo/libjpeg-turbo#764]
**Have you searched the existing issues (both open and closed) in the libjpeg-turbo issue tracker to ensure that this bug report is not a duplicate?**

Yes

**Does this bug report describe one of the [two known and unsolvable issues with the JPEG format](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf)?**

No

**Clear and concise description of the bug:**

`save_marker` uses an O(n) linked-list walk to save a marker: https://github.com/libjpeg-turbo/libjpeg-turbo/blob/7fa4b5b762c9a99b46b0b7838f5fd55071b92ea5/jdmarker.c#L826

There does not appear to be an upper limit on the number of markers in a JPEG.

Thus, we experience an O(n^2) performance hit if a maliciously-crafted JPEG includes millions of meaningless markers.

**Steps to reproduce the bug (using *only* libjpeg-turbo):**

View malicious JPEG containing millions of markers; process hang will occur.

**Image(s) needed in order to reproduce the bug (if applicable):**

Will provide upon request, probably best not to include here as it will hang a browser.

**Expected behavior:**

Marker saving should be implemented in an O(1) fashion.

**Observed behavior:**

O(n^2) slowdown with unlimited upper bound for n

**Platform(s) (compiler version, operating system version, CPU) on which the bug was observed:**

Linux + oss-fuzz

**libjpeg-turbo release(s), commit(s), or branch(es) in which the bug was observed (always test the tip of the main branch or the latest [stable pre-release](https://libjpeg-turbo.org/DeveloperInfo/PreReleases) to verify that the bug hasn't already been fixed):**

Latest

**If the bug is a regression, the specific commit that introduced the regression (use `git bisect` to determine this):**

n/a

**Additional information:**

n/a

**Comment by dcommander** [comment]
Reproducible using `jpegtran -copy all` or `tj3Transform()` (if `TJXOPT_COPYNONE` is not specified.)  Not reproducible with djpeg or `tj3Decompress*()`.

As far as I know, there is no specified limit on the number of markers, just like there is no specified limit on the number of scans.  So this is a similar exploit to LJT­-01-­003 described [here](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf).  The way to mitigate such issues is to make libjpeg warnings fatal.  (See the `-strict` switch in **cjpeg.c** and **djpeg.c** for examples.  In the TurboJPEG API, you can also simply set `TJPARAM_STOPONWARNING`.)  If you do that, then the image is almost immediately rejected with the error "Corrupt JPEG data: 6 extraneous bytes before marker 0xe1".  Historically, the libjpeg API did not treat warnings as fatal, so our default behavior is conformant to that legacy behavior.  (Reference my frequent rant about the fact that we are still beholden to a 33-year-old API and architecture that was designed around the limitations of MS-DOS machines.)  The tradeoff is that, if warnings are non-fatal (default behavior), then an application can continue to decode as much of a corrupt JPEG image as possible while rejecting only the corrupted segments.  However, treating warnings as non-fatal also opens up the application to exploits such as this one and the one described in the aforementioned report.  Thus, it is strongly recommended (and has been strongly recommended for years) that browsers and other applications that must process arbitrary JPEG images treat libjpeg warnings as fatal.  The worst case in Chrome, however, appears to be a frozen tab, not freezing the whole browser.

I don't understand what you mean by "Marker saving should be implemented in an O(1) fashion".  How do I do that if there is no specified limit on the number of markers?  I can't just impose my own arbitrary limit.

**Comment by johnstiles-google** [comment]
Sorry, I meant O(1) per marker, which would be O(n) overall. This could be
accomplished by remembering the list end (akin to std::list) or by adding
new markers to the front of the list and reversing it later.

**Comment by johnstiles-google** [comment]
Re treating warnings as fatal, I haven't looked into the data in this JPEG in much detail, but it seems like this might be a red herring? Wouldn't it still be possible to generate a JPEG file with no warnings that still contained millions of markers? 

I can look into the impact of enabling fatal warnings in Chrome, but that seems like it would be orthogonal to this issue. It runs the real risk of preventing Chrome users from viewing existing content that contains benign warnings, so I think it might be a difficult decision. In a surface as large as "the entire Internet" there will be a large number of images which are simultaneously corrupt, viewable in practice, and important to someone.

**Comment by dcommander** [comment]
> Re treating warnings as fatal, I haven't looked into the data in this JPEG in much detail, but it seems like this might be a red herring? Wouldn't it still be possible to generate a JPEG file with no warnings that still contained millions of markers?

Maybe.  It was possible to generate a correct JPEG file that produced no libjpeg warnings but still triggered the aforementioned LJT-01-003 issue.  However, that file was much larger in size than the malformed file that the security researchers used, so it triggered about 1 minute of compute time per megabyte rather than 10 minutes per megabyte.  I suspect that that would be the case here as well, that the issue would still exist but would be greatly diminished in severity and scope if libjpeg warnings were fatal.  The nature of image compression means that compute time will always be somewhat disproportionate to I/O time, i.e. disproportionate to the image dimensions and file size.  Thus, it will always be possible to trigger a large amount of compute time if the file size or image dimensions are large enough.  Really, the crux of these complaints is that the compute time is grossly disproportionate to the I/O time, i.e. that it is a clear outlier among JPEG images with the same file size.

> I can look into the impact of enabling fatal warnings in Chrome, but that seems like it would be orthogonal to this issue. It runs the real risk of preventing Chrome users from viewing existing content that contains benign warnings, so I think it might be a difficult decision. In a surface as large as "the entire Internet" there will be a large number of images which are simultaneously corrupt, viewable in practice, and important to someone.

Why is Chrome calling `jpeg_save_markers()` in the first place?  Applications wouldn't normally do that when decompressing a JPEG image, unless they were saving the ICC profile (which does not trigger this issue, because the reproducer only has APP1 markers.)  I'm sure it's possible to craft a reproducer that uses APP2 markers instead.  I'm just trying to understand whether calling `jpeg_save_markers()` is something Chrome does all the time or whether it only does so under special circumstances, and whether it would be possible to make warnings fatal only under those special circumstances.

Otherwise, yes, what you describe is exactly the tradeoff of making libjpeg warnings fatal, and it's the reason why they aren't fatal by default.  The other way that security-critical applications work around LJT-01-003 is by limiting the number of progressive scans that they will process.  (It may be possible to justify, for instance, a progressive JPEG image with 100 scans, but it's really hard to justify 1000 scans.)  However, there isn't any equivalent way to limit the number of markers except by ignoring markers of a certain type.

There is a limit to what I can do as long as I am still constrained by a 33-year-old API and architecture.  However, I will at least look into the possibility of optimizing the linked list algorithm.

**Comment by johnstiles-google** [comment]
> Why is Chrome calling jpeg_save_markers() in the first place?

We are not calling it; we are calling `jpeg_read_header`. As part of reading the header, libjpeg is storing a linked list of markers automatically.

```
save_marker
read_markers
consume_markers
chromium_jpeg_consume_input
chromium_jpeg_read_header
SkJpegCodec::ReadHeader
```

> Really, the crux of these complaints is that the compute time is grossly disproportionate to the I/O time, i.e. that it is a clear outlier among JPEG images with the same file size.

My concern is that compute time increases in a superlinear fashion with no bound. Looking at the description of the unnecessary-scans issue, it is described as "the decode time goes up linearly with
the number of 100­byte scans one adds to the image." I have not looked at it in great detail, but if adding a marker incurs an O(n) scan over the list of every existing marker, and the marker list is unbounded, then the time goes up _exponentially_ as markers are added. 

It seems as if this problem could be entirely resolved by remembering the end of the marker list, or using some other representation; this would bring us to linear time and the cost of adding an individual marker will be extremely low. This doesn't seem to have the same implicit challenges as LJT­-01-­003, which legitimately _needs_ that CPU time.

**Comment by dcommander** [comment]
> > Why is Chrome calling jpeg_save_markers() in the first place?
> 
> We are not calling it; we are calling `jpeg_read_header`. As part of reading the header, libjpeg is storing a linked list of markers automatically.
> 
> ```
> save_marker
> read_markers
> consume_markers
> chromium_jpeg_consume_input
> chromium_jpeg_read_header
> SkJpegCodec::ReadHeader
> ```

I wonder why the issue can't be reproduced with djpeg, then.

**Comment by johnstiles-google** [comment]
Oh, my apologies, I was mistaken. We do call `jpeg_save_markers` to keep the color management markers. 

https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/src/codec/SkJpegCodec.cpp;l=95-102;drc=2319f1ae8fe42525f8b6a1969a1cee67bdbee290

**Comment by dcommander** [comment]
OK, that makes sense.  The reproducer contains APP1 markers, which is what EXIF uses.

I think I have a solution for storing the end of the linked list.  I need to do some more testing, though, so it'll be tomorrow before I can push it.

**Comment by johnstiles-google** [comment]
That's great news, thank you for the quick turnaround!

**Comment by kleisauke** [comment]
FWIW, I think the testcase in https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=24383 might also be relevant here. libvips currently workarounds this by treating 100+ warnings as fatal, see commit: https://github.com/libvips/libvips/commit/df26bd1e46c67ed7745996627ca0d83e5a6d678d.

One could still reproduce that by passing the `unlimited` flag.
```console
$ time vips copy clusterfuzz-testcase-minimized-smartcrop_fuzzer-5155920259317760[unlimited] x.jpg

(vips:709796): VIPS-WARNING **: 07:35:45.468: read gave 72738 warnings

(vips:709796): VIPS-WARNING **: 07:35:45.468: VipsJpeg: Corrupt JPEG data: 2 extraneous bytes before marker 0xed
VipsJpeg: Corrupt JPEG data: 2 extraneous bytes before marker 0xed
VipsJpeg: Quantization table 0x01 was not defined


real    0m26.938s
user    0m26.307s
sys     0m0.055s
```

**Comment by johnstiles-google** [comment]
Great! Thank you very much. I'll try to get this rolled into Chromium this week.

**Comment by dcommander** [comment]
I just released 3.0.3, so it will be a few months before 3.0.4 lands.
